qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,349 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Dict, Tuple
+
+import attrs
+import numpy as np
+from numpy.typing import ArrayLike
+
+from qualtran import Bloq, BloqBuilder, SelectionRegister, Signature, SoquetT
+from qualtran.bloqs.arithmetic import Add
+from qualtran.bloqs.basic_gates import OneEffect, OneState, ZeroEffect, ZeroState
+from qualtran.bloqs.basic_gates.rotation import Rx
+from qualtran.bloqs.qrom import QROM
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+
```

Comment:
The problem with the paper is that it is not really about state preparation, but mentions it and gives the main idea behind the algorithm (the best I could find). Many of the things that are used in the code are not really explained there but referenced to other huge papers where it is nowhere to be found. So I had to figure them out myself or with the help of @StefanoPolla. I will add an outline of what it does though, because as it is it might be a bit hard to understand.

Reply:
To get a short explanation we could connect the QROM call to the first line of Eq.8 (loading the binary representation of $\theta$ depending on $|y\rangle$), the controlled-rotation to the second line, and the QROM computation to the third.

The additional control needs a mention but no special explanation.